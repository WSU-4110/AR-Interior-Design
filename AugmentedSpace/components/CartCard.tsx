import { Image, ImageSourcePropType, Pressable, Text, View } from "react-native";
import { useTheme } from "@react-navigation/native";
import { useEffect, useState } from "react";
import { getDownloadURL, getStorage, ref } from "firebase/storage";
import { getAuth } from "firebase/auth";
type CartCardProps = {
  itemName: string;
  brandName: string;
  itemCost: number;
  imagePath: string;
  className?: string;
  onPress?: () => void;
  onRemove: () => void;
};

export default function CardCard(props: CartCardProps) {
  const { colors } = useTheme();
  const [imageUrl, setImageUrl] = useState("");
  const { currentUser } = getAuth();

  useEffect(() => {
    //retrieve image url from firebase
    const getImageUrl = async () => {
      const storage = getStorage();
      const reference = ref(storage, props.imagePath);
      await getDownloadURL(reference).then((x) => {
        setImageUrl(x);
      })
    }

    getImageUrl();
  }, []);

  return (
    <Pressable
      testID="cart-card-pressable"
      className="w-1/2 h-auto flex-1 mx-1 p-2 rounded-2xl"
      onPress={props.onPress}
      style={{
        backgroundColor: colors.card,
        shadowOffset: { width: 2, height: 2 },
        shadowColor: colors.shadow,
        shadowOpacity: 1,
      }}
    >
      <Image
        testID="cart-card-image"
        className="self-center h-32 w-32 my-2"
        source={{uri: imageUrl}}
      />
      <View style={{
        flexDirection: "row",
        justifyContent: "space-evenly",
        alignItems: "center",
        marginVertical: 80,
      }}
        className="m-1">
        <Text className="italic font-semibold" style={{ color: colors.text }}>
          {props.brandName}
        </Text>
        <Text className="italic text-xl" style={{ color: colors.text }}>
          {props.itemName}
        </Text>
        
        <Text className="font-semibold" style={{ color: colors.text }}>
          ${props.itemCost}
        </Text>
        
      </View>
      <Pressable onPress={props.onRemove} style = {{
            padding: 10,
            alignItems: 'center',
            borderTopWidth: 3,
            borderColor: colors.border,
        }}>
            <Text style={{color: colors.text, fontWeight: 'bold',}}>Remove</Text>
        </Pressable>
    </Pressable>
  );
}
