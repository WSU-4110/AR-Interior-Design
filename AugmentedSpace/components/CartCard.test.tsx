import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import CartCard from './CartCard'; // Adjust the import path as necessary

// Mocking the necessary hooks and methods from Firebase and @react-navigation/native
jest.mock('firebase/auth', () => ({
  getAuth: jest.fn().mockReturnValue({
    currentUser: { uid: 'test-user-id' },
  }),
}));

jest.mock('firebase/storage', () => ({
  getStorage: jest.fn(),
  ref: jest.fn(),
  getDownloadURL: jest.fn(() => Promise.resolve('image-test-url')),
}));

jest.mock('@react-navigation/native', () => ({
  useTheme: jest.fn().mockReturnValue({
    colors: {
      card: 'white',
      text: 'black',
      shadow: 'grey',
      border: 'blue',
    },
  }),
}));

describe('<CartCard />', () => {
  const mockOnRemove = jest.fn();
  const mockOnPress = jest.fn();

  beforeEach(() => {
    // Reset the mock calls before each test
    mockOnRemove.mockClear();
    mockOnPress.mockClear();
  });

  it('renders correctly', () => {
    const { toJSON } = render(<CartCard
      itemName="Test Item"
      brandName="Test Brand"
      itemCost={9.99}
      imagePath="path/to/image.jpg"
      onRemove={mockOnRemove}
      onPress={mockOnPress}
    />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('displays the correct itemName, brandName, and itemCost', () => {
    const { getByText } = render(<CartCard
      itemName="Unique Item Name"
      brandName="Unique Brand Name"
      itemCost={123.45}
      imagePath="path/to/image.jpg"
      onRemove={mockOnRemove}
      onPress={mockOnPress}
    />);
    
    expect(getByText('Unique Brand Name')).toBeTruthy();
    expect(getByText('Unique Item Name')).toBeTruthy();
    expect(getByText('$123.45')).toBeTruthy();
  });

  it('calls onPress when the card is pressed', () => {
    const { getByTestId } = render(<CartCard
      itemName="Test Item"
      brandName="Test Brand"
      itemCost={9.99}
      imagePath="path/to/image.jpg"
      onRemove={mockOnRemove}
      onPress={mockOnPress}
    />);

    // Assuming you assign a testID to your Pressable component
    const pressable = getByTestId('cart-card-pressable');
    fireEvent.press(pressable);
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it('calls onRemove when the remove button is pressed', () => {
    const { getByText } = render(<CartCard
      itemName="Test Item"
      brandName="Test Brand"
      itemCost={9.99}
      imagePath="path/to/image.jpg"
      onRemove={mockOnRemove}
      onPress={mockOnPress}
    />);
    
    const removeButton = getByText('Remove');
    fireEvent.press(removeButton);
    expect(mockOnRemove).toHaveBeenCalledTimes(1);
  });
});
