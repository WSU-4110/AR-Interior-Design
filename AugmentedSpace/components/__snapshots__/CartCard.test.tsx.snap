// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCard /> renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "width": "50%",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "marginLeft": 4,
        "marginRight": 4,
      },
      {
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
      },
      {
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "borderTopLeftRadius": 16,
        "borderTopRightRadius": 16,
      },
      {
        "backgroundColor": "white",
        "shadowColor": "grey",
        "shadowOffset": {
          "height": 2,
          "width": 2,
        },
        "shadowOpacity": 1,
      },
    ]
  }
  testID="cart-card-pressable"
>
  <Image
    source={
      {
        "uri": "",
      }
    }
    style={
      [
        {
          "alignSelf": "center",
        },
        {
          "height": 128,
        },
        {
          "width": 128,
        },
        {
          "marginBottom": 8,
          "marginTop": 8,
        },
      ]
    }
    testID="cart-card-image"
  />
  <View
    style={
      [
        {
          "marginBottom": 4,
          "marginLeft": 4,
          "marginRight": 4,
          "marginTop": 4,
        },
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginVertical": 80,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontStyle": "italic",
          },
          {
            "fontWeight": "600",
          },
          {
            "color": "black",
          },
        ]
      }
    >
      Test Brand
    </Text>
    <Text
      style={
        [
          {
            "fontStyle": "italic",
          },
          {
            "fontSize": 20,
            "lineHeight": 28,
          },
          {
            "color": "black",
          },
        ]
      }
    >
      Test Item
    </Text>
    <Text
      style={
        [
          {
            "fontWeight": "600",
          },
          {
            "color": "black",
          },
        ]
      }
    >
      $
      9.99
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderColor": "blue",
        "borderTopWidth": 3,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontWeight": "bold",
        }
      }
    >
      Remove
    </Text>
  </View>
</View>
`;
